File: backend/express/src/db/seed.ts
import {db} from "./db.ts";
import {
  users,
  teams,
  projects,
  tasks,
  teamMembers,
  projectMembers,
  starredTasks,
  notifications,
  userNotifications,
  userSettings,
  activityHistory,
  taskComments,
  UserRole,
} from "./schema.ts";
import bcrypt from "bcrypt";
import crypto from "crypto";

async function hashPassword(password: string) {
  const hashedPassword = await bcrypt.hash(password, 10);
  return hashedPassword;
}

export async function seed() {
  try {
    console.log("Checking if seeding is necessary...");

    const existingUsers = await db.select().from(users).limit(1);
    if (existingUsers.length > 0) {
      console.log("Database already seeded. Skipping...");
      return;
    }

    console.log("Seeding started...");

    const usersData = [
      {
        email: "admin@example.com",
        password: await hashPassword("123456789"),
        name: "Admin User",
        role: UserRole.ADMIN,
        isVerified: true,
        createdAt: new Date(),
      },
      {
        email: "manager@example.com",
        password: await hashPassword("123456789"),
        name: "Project Manager",
        role: UserRole.USER,
        isVerified: true,
        createdAt: new Date(),
      },
      {
        email: "dev1@example.com",
        password: await hashPassword("123456789"),
        name: "Developer One",
        role: UserRole.USER,
        isVerified: true,
        createdAt: new Date(),
      },
      {
        email: "dev2@example.com",
        password: await hashPassword("123456789"),
        name: "Developer Two",
        role: UserRole.USER,
        isVerified: true,
        createdAt: new Date(),
      },
      {
        email: "tester@example.com",
        password: await hashPassword("123456789"),
        name: "QA Tester",
        role: UserRole.USER,
        isVerified: true,
        createdAt: new Date(),
      },
    ];

    await db.insert(users).values(usersData);
    console.log("Users seeded.");

    await db.insert(userSettings).values(
      usersData.map((user) => ({
        userEmail: user.email,
        emailDigest: true,
        taskReminders: true,
      }))
    );
    console.log("User settings seeded.");

    const teamsData = [
      {id: 1, name: "Frontend Team", createdAt: new Date()},
      {id: 2, name: "Backend Team", createdAt: new Date()},
      {id: 3, name: "DevOps Team", createdAt: new Date()},
      {id: 4, name: "QA Team", createdAt: new Date()},
    ];

    await db.insert(teams).values(teamsData);
    console.log("Teams seeded.");

    const teamMembersData = [
      {teamId: 1, email: "admin@example.com"},
      {teamId: 2, email: "admin@example.com"},
      {teamId: 3, email: "admin@example.com"},
      {teamId: 4, email: "admin@example.com"},

      {teamId: 1, email: "manager@example.com"},
      {teamId: 2, email: "manager@example.com"},

      {teamId: 1, email: "dev1@example.com"},

      {teamId: 2, email: "dev2@example.com"},

      {teamId: 4, email: "tester@example.com"},
    ];

    await db.insert(teamMembers).values(teamMembersData);
    console.log("Team members seeded.");

    const projectsData = [
      {
        id: "proj-fe-1",
        name: "Website Redesign",
        teamId: 1,
        createdAt: new Date(),
      },
      {
        id: "proj-be-1",
        name: "API Development",
        teamId: 2,
        createdAt: new Date(),
      },
      {
        id: "proj-qa-1",
        name: "Testing Framework",
        teamId: 4,
        createdAt: new Date(),
      },
    ];

    await db.insert(projects).values(projectsData);
    console.log("Projects seeded.");

    const projectMembersData = [
      {projectId: "proj-fe-1", email: "admin@example.com"},
      {projectId: "proj-fe-1", email: "manager@example.com"},
      {projectId: "proj-fe-1", email: "dev1@example.com"},

      {projectId: "proj-be-1", email: "admin@example.com"},
      {projectId: "proj-be-1", email: "manager@example.com"},
      {projectId: "proj-be-1", email: "dev2@example.com"},

      {projectId: "proj-qa-1", email: "admin@example.com"},
      {projectId: "proj-qa-1", email: "tester@example.com"},
    ];

    await db.insert(projectMembers).values(projectMembersData);
    console.log("Project members seeded.");

    const tasksData = [
      {
        id: "task-1",
        name: "Design Homepage",
        description: "Create new homepage design",
        projectId: "proj-fe-1",
        createdAt: new Date(),
        stars: 2,
      },
      {
        id: "task-2",
        name: "Implement Auth API",
        description: "Develop authentication endpoints",
        projectId: "proj-be-1",
        createdAt: new Date(),
        stars: 1,
      },
      {
        id: "task-3",
        name: "Write E2E Tests",
        description: "Create end-to-end tests for main features",
        projectId: "proj-qa-1",
        createdAt: new Date(),
        stars: 0,
      },
    ];

    await db.insert(tasks).values(tasksData);
    console.log("Tasks seeded.");

    const taskCommentsData = [
      {
        id: crypto.randomBytes(16).toString("hex"),
        taskId: "task-1",
        userId: "dev1@example.com",
        content: "Started working on the mockups",
        createdAt: new Date(),
      },
      {
        id: crypto.randomBytes(16).toString("hex"),
        taskId: "task-2",
        userId: "dev2@example.com",
        content: "JWT implementation completed",
        createdAt: new Date(),
      },
    ];

    await db.insert(taskComments).values(taskCommentsData);
    console.log("Task comments seeded.");

    const starredTasksData = [
      {
        userId: "manager@example.com",
        taskId: "task-1",
        createdAt: new Date(),
      },
      {
        userId: "dev1@example.com",
        taskId: "task-1",
        createdAt: new Date(),
      },
      {
        userId: "dev2@example.com",
        taskId: "task-2",
        createdAt: new Date(),
      },
    ];

    await db.insert(starredTasks).values(starredTasksData);
    console.log("Starred tasks seeded.");

    const notificationsData = [
      {
        id: "notif-1",
        type: "task_assignment",
        title: "New Task Assigned",
        content: "You have been assigned to 'Design Homepage'",
        createdAt: new Date(),
        link: "/tasks/task-1",
      },
      {
        id: "notif-2",
        type: "comment",
        title: "New Comment",
        content: "New comment on 'Implement Auth API'",
        createdAt: new Date(),
        link: "/tasks/task-2",
      },
    ];

    await db.insert(notifications).values(notificationsData);
    console.log("Notifications seeded.");

    const userNotificationsData = [
      {
        userId: "dev1@example.com",
        notificationId: "notif-1",
      },
      {
        userId: "dev2@example.com",
        notificationId: "notif-2",
      },
    ];

    await db.insert(userNotifications).values(userNotificationsData);
    console.log("User notifications seeded.");

    const activityHistoryData = [
      {
        id: crypto.randomBytes(16).toString("hex"),
        userId: "admin@example.com",
        description: "Created Frontend Team",
        doneAt: new Date(),
      },
      {
        id: crypto.randomBytes(16).toString("hex"),
        userId: "manager@example.com",
        description: "Created Website Redesign project",
        doneAt: new Date(),
      },
      {
        id: crypto.randomBytes(16).toString("hex"),
        userId: "dev1@example.com",
        description: "Added comment to Design Homepage task",
        doneAt: new Date(),
      },
    ];

    await db.insert(activityHistory).values(activityHistoryData);
    console.log("Activity history seeded.");

    console.log("Seeding completed successfully.");
  } catch (err) {
    console.error("Error seeding database:", err);
  }
}

File: backend/golang/db/seed.go
package db

import (
	"log"
	"time"

	"github.com/google/uuid"
	"github.com/maiko26/multi-backend-app/backend/golang/internal/model"
	"golang.org/x/crypto/bcrypt"
)

func Seed(database *DB) {
	var userCount int64
	database.DB.Model(&model.User{}).Count(&userCount)
	if userCount > 0 {
		log.Println("Database already seeded, skipping...")
		return
	}

	log.Println("Seeding database...")

	hashedPassword, err := bcrypt.GenerateFromPassword([]byte("123456789"), 10)
	if err != nil {
		log.Fatalf("Failed to hash password: %v", err)
	}
	passwordStr := string(hashedPassword)

	users := []model.User{
		{Email: "admin@example.com", Password: passwordStr, Name: "Admin User", Role: "admin", IsVerified: true},
		{Email: "alice@example.com", Password: passwordStr, Name: "Alice Smith", Role: "user", IsVerified: true},
		{Email: "bob@example.com", Password: passwordStr, Name: "Bob Johnson", Role: "user", IsVerified: true},
		{Email: "charlie@example.com", Password: passwordStr, Name: "Charlie Brown", Role: "user", IsVerified: true},
		{Email: "diana@example.com", Password: passwordStr, Name: "Diana Prince", Role: "user", IsVerified: true},
	}
	for i := range users {
		users[i].EmailNotifications = true
		users[i].PushNotifications = true
		users[i].SubscriptionStatus = "free"
	}
	database.DB.Create(&users)

	settings := []model.UserSettings{
		{UserEmail: "admin@example.com", EmailDigest: true, TaskReminders: true},
		{UserEmail: "alice@example.com", EmailDigest: true, TaskReminders: true},
		{UserEmail: "bob@example.com", EmailDigest: true, TaskReminders: true},
		{UserEmail: "charlie@example.com", EmailDigest: true, TaskReminders: true},
		{UserEmail: "diana@example.com", EmailDigest: true, TaskReminders: true},
	}
	database.DB.Create(&settings)

	teams := []model.Team{
		{Name: "Frontend"},
		{Name: "Backend"},
		{Name: "DevOps"},
		{Name: "QA"},
	}
	database.DB.Create(&teams)

	teamMembers := []model.TeamMember{
		{TeamID: teams[0].ID, Email: "alice@example.com"},
		{TeamID: teams[0].ID, Email: "bob@example.com"},
		{TeamID: teams[1].ID, Email: "charlie@example.com"},
		{TeamID: teams[1].ID, Email: "diana@example.com"},
		{TeamID: teams[2].ID, Email: "bob@example.com"},
		{TeamID: teams[3].ID, Email: "alice@example.com"},
		{TeamID: teams[3].ID, Email: "charlie@example.com"},
	}
	database.DB.Create(&teamMembers)

	projects := []model.Project{
		{ID: uuid.New().String()[:8], Name: "Website Redesign", TeamID: teams[0].ID},
		{ID: uuid.New().String()[:8], Name: "API Development", TeamID: teams[1].ID},
		{ID: uuid.New().String()[:8], Name: "Testing Framework", TeamID: teams[3].ID},
	}
	database.DB.Create(&projects)

	tasks := []model.Task{
		{ID: uuid.New().String()[:8], Name: "Design homepage", Description: "Create wireframes and mockups", ProjectID: projects[0].ID},
		{ID: uuid.New().String()[:8], Name: "Implement REST API", Description: "Build CRUD endpoints", ProjectID: projects[1].ID, Finished: true, FinishedBy: "charlie@example.com"},
		{ID: uuid.New().String()[:8], Name: "Write unit tests", Description: "Cover core services", ProjectID: projects[2].ID},
	}
	database.DB.Create(&tasks)

	comments := []model.TaskComment{
		{ID: uuid.New().String()[:8], TaskID: tasks[0].ID, UserID: "alice@example.com", Content: "Looking great!"},
		{ID: uuid.New().String()[:8], TaskID: tasks[1].ID, UserID: "charlie@example.com", Content: "API is ready for review"},
	}
	database.DB.Create(&comments)

	starredTasks := []model.StarredTask{
		{UserID: "alice@example.com", TaskID: tasks[0].ID},
		{UserID: "bob@example.com", TaskID: tasks[1].ID},
	}
	database.DB.Create(&starredTasks)

	projectMembers := []model.ProjectMember{
		{ProjectID: projects[0].ID, Email: "alice@example.com"},
		{ProjectID: projects[0].ID, Email: "bob@example.com"},
		{ProjectID: projects[1].ID, Email: "charlie@example.com"},
		{ProjectID: projects[1].ID, Email: "diana@example.com"},
		{ProjectID: projects[2].ID, Email: "alice@example.com"},
	}
	database.DB.Create(&projectMembers)

	notifications := []model.Notification{
		{ID: uuid.New().String()[:8], Type: "task_assignment", Title: "New Task", Content: "You have been assigned a new task", Link: "/tasks"},
		{ID: uuid.New().String()[:8], Type: "comment", Title: "New Comment", Content: "Someone commented on your task", Link: "/tasks"},
	}
	database.DB.Create(&notifications)

	userNotifications := []model.UserNotification{
		{UserID: "alice@example.com", NotificationID: notifications[0].ID},
		{UserID: "bob@example.com", NotificationID: notifications[1].ID},
	}
	database.DB.Create(&userNotifications)

	activities := []model.ActivityHistory{
		{ID: uuid.New().String()[:8], UserID: "admin@example.com", Description: "Created new project"},
		{ID: uuid.New().String()[:8], UserID: "alice@example.com", Description: "Completed task: Design homepage"},
		{ID: uuid.New().String()[:8], UserID: "charlie@example.com", Description: "Added comment to task"},
	}
	database.DB.Create(&activities)

	chatRooms := []model.ChatRoom{
		{RoomID: "general"},
	}
	database.DB.Create(&chatRooms)

	now := time.Now()
	messages := []model.ChatMessage{
		{RoomID: "general", SenderID: "alice@example.com", Content: "Hey everyone!", CreatedAt: now},
		{RoomID: "general", SenderID: "bob@example.com", Content: "Hi Alice!", CreatedAt: now.Add(time.Minute)},
	}
	database.DB.Create(&messages)

	log.Println("Database seeded successfully!")
}
